### PR TITLE
support base64 secret for jwt auth

### LINKED ISSUE #1223: Enhancement: jwt-auth should support base64 encoded secret
The secret in jwt-auth plugin is now only string typed, this is not secure enough:

- The string contains only printable characters which makes the secret's possible value size much smaller than byte array with the same size.
- It's easy to generate secure random byte array, but not so easy to generate a secure random string. If we generate a secret without enough randomness, the secret is much weaker and not compatible with jwt standards(Although JWT standard didn't mention too much about how the key should be generated, the standards it referenced did state the secret should be random enough to protect the signature).

Ref:
JWT: https://tools.ietf.org/html/rfc7519
HMAC: https://tools.ietf.org/html/rfc2104

### PR DESCRIPTION
**Pull request by paoying**
NOTE: Please read the Contributing.md guidelines before submitting your patch:

https://github.com/apache/incubator-apisix/blob/master/Contributing.md#how-to-add-a-new-feature-or-change-an-existing-one

### Summary

Added base64 encoded secret support to jwt-auth plugin

### Full changelog

* lua/apisix/plugins/jwt-auth.lua
* t/plugin/jwt-auth.t

### Issues resolved

Fix #1223

### CODE REVIEW

File: lua/apisix/plugins/jwt-auth.lua [hunk 1]
```diff
@@ -75,7 +82,11 @@ function _M.check_schema(conf)
     end
 
     if not conf.secret then
-        conf.secret = core.id.gen_uuid_v4()
+        if base64_secret then
+            conf.secret = ngx_encode_base64(resty_random.bytes(32))
```
Comment (moonming) [inline lua/apisix/plugins/jwt-auth.lua]: We should uniformly generate the random number algorithm, or only specified by the user, change the `secret` to a required field

Comment (paoying) [reply lua/apisix/plugins/jwt-auth.lua]: changed to use rest_random to generate ascii secret

File: lua/apisix/plugins/jwt-auth.lua [hunk 2]
```diff
@@ -75,7 +82,7 @@ function _M.check_schema(conf)
     end
 
     if not conf.secret then
-        conf.secret = core.id.gen_uuid_v4()
```
Comment (moonming) [inline lua/apisix/plugins/jwt-auth.lua]: we can remove lua-resty-uuid now

File: lua/apisix/plugins/jwt-auth.lua [hunk 3]
```diff
@@ -75,7 +82,7 @@ function _M.check_schema(conf)
...
+        conf.secret = ngx_encode_base64(resty_random.bytes(32, true))
```
Comment (moonming) [inline lua/apisix/plugins/jwt-auth.lua]: need more test cases, for example:
- no `secret` and no `base64_secret`
- has 'secret' and no `base64_secret`
etc.

File: t/plugin/jwt-auth.t [hunk 4]
```diff
@@ -42,7 +42,7 @@ __DATA__
 --- request
 GET /t
 --- response_body_like eval
-qr/{"algorithm":"HS256","secret":"\w+-\w+-\w+-\w+-\w+","exp":86400}/
+qr/{"algorithm":"HS256","secret":".+","exp":86400}/
```
Comment (moonming) [inline t/plugin/jwt-auth.t]: need check is base64 format.

File: lua/apisix/plugins/jwt-auth.lua (hunk 3 above)
Comment (moonming) [reply lua/apisix/plugins/jwt-auth.lua]: need doc for this PR too

File: lua/apisix/plugins/jwt-auth.lua [hunk 5]
```diff
@@ -35,6 +38,10 @@ local schema = {
             enum = {"HS256", "HS384", "HS512", "RS256", "ES256"}
         },
         exp = {type = "integer", minimum = 1},
+        base64_secret = {
+                    type = "boolean",
```
Comment (membphis) [inline lua/apisix/plugins/jwt-auth.lua]: bad indentation

File: lua/apisix/plugins/jwt-auth.lua [hunk 6]
```diff
@@ -114,6 +121,13 @@ local function fetch_jwt_token(ctx)
     return val, err
 end
 
+local function get_secret(conf)
+    if conf.base64_secret then
```
Comment (membphis) [inline lua/apisix/plugins/jwt-auth.lua]:
```lua
if ... then
end

...
return
```

File: t/plugin/jwt-auth.t [hunk 7]
```diff
@@ -327,3 +327,118 @@ code: true body: passed
 code: true body: passed
 --- no_error_log
 [error]
+
```
Comment (membphis) [inline t/plugin/jwt-auth.t]: your test case style is wrong.

please take a look at this doc.

https://github.com/apache/incubator-apisix/blob/429f0165c202035b18b4437a84c6edc9007fc440/Contributing.md#check-code-style-and-test-case-style

**Comment by moonming** [comment]
@paoying any update?

**Comment by paoying** [comment]
> @paoying any update?

Got quite busy these days, might need more time.

**Comment by moonming** [comment]
@paoying any update?

File: lua/apisix/plugins/jwt-auth.lua (hunk 3 above)
Comment (paoying) [reply lua/apisix/plugins/jwt-auth.lua:85]: > need more test cases, for example:
> 
> * no `secret` and no `base64_secret`
> * has 'secret' and no `base64_secret`
>   etc.

don't get this, base64_secret is a boolean flag, it's not another secret.

Comment (paoying) [reply lua/apisix/plugins/jwt-auth.lua:85]: > need doc for this PR too

do you mean need to change jwt-auth doc?

File: apisix/plugins/jwt-auth.lua [hunk 8]
```diff
@@ -35,6 +38,10 @@ local schema = {
             enum = {"HS256", "HS384", "HS512", "RS256", "ES256"}
         },
         exp = {type = "integer", minimum = 1},
+        base64_secret = {
```
Comment (membphis) [inline apisix/plugins/jwt-auth.lua:41]: we need to check `secret` value if `base64_secret` is true.

It should be a valid base64 encoded string.

Comment (moonming) [reply apisix/plugins/jwt-auth.lua:41]: I think we can merge fisrt, then add schema check, Chinesse doc and more test cases in another PR.

Comment (membphis) [reply apisix/plugins/jwt-auth.lua:41]: agree +1

**Comment by membphis** [comment]
merged, many thx @paoying

**Pull request merged**